### PR TITLE
Feature/optimize closed line check

### PR DESCRIPTION
**Pull request by ivandrofly**
More efficient sentence ending check for both memory and speed plus method specific unit-test

### CODE REVIEW

File: src/libse/Common/StringExtensions.cs [hunk 1]
```diff
@@ -757,24 +757,70 @@ public static bool HasSentenceEnding(this string value)
...
+
         public static bool HasSentenceEnding(this string value, string twoLetterLanguageCode)
         {
             if (string.IsNullOrEmpty(value))
             {
-                return false;
+                return true;
             }
 
-            var s = HtmlUtil.RemoveHtmlTags(value, true).TrimEnd('"').TrimEnd('”');
```
Comment (ivandrofly) [inline src/libse/Common/StringExtensions.cs:767]: This was the main performance killer!

File: src/libse/Common/StringExtensions.cs [hunk 2]
```diff
@@ -757,24 +757,70 @@ public static bool HasSentenceEnding(this string value)
...
+            }
+
+            // value contains only quotes
+            if (checkIndex < 0)
             {
                 return false;
             }
 
-            var last = s[s.Length - 1];
-            return last == '.' || last == '!' || last == '?' || last == ']' || last == ')' || last == '…' || last == '♪' || last == '؟' ||
```
Comment (ivandrofly) [inline src/libse/Common/StringExtensions.cs:774]: A little complex too many logical ||

File: src/ui/Logic/Ocr/OcrFixEngine.cs [hunk 3]
```diff
@@ -667,12 +667,11 @@ private string FixWordViaReplaceList(string word)
         public static string FixFrenchLApostrophe(string input, string tag, string lastLine)
         {
             var text = input;
-            var endingBeforeThis = string.IsNullOrEmpty(lastLine) || lastLine.EndsWith('.') || lastLine.EndsWith('!') || lastLine.EndsWith('?') ||
-                                   lastLine.EndsWith(".</i>", StringComparison.Ordinal) || lastLine.EndsWith("!</i>", StringComparison.Ordinal) || lastLine.EndsWith("?</i>", StringComparison.Ordinal) ||
-                                   lastLine.EndsWith(".</font>", StringComparison.Ordinal) || lastLine.EndsWith("!</font>", StringComparison.Ordinal) || lastLine.EndsWith("?</font>", StringComparison.Ordinal);
+            var isPreviousLineClose = lastLine.HasSentenceEnding();
```
Comment (ivandrofly) [inline src/ui/Logic/Ocr/OcrFixEngine.cs:670]: Use already defined method to check for sentence ending

**Comment by niksedk** [comment]
Cool !

Null or empty should probably return `false`.

Could you also make it work with ASSA tags, like `"{\\i1}How are you?{\\i0}"` ? :)

**Comment by ivandrofly** [comment]
> Null or empty should probably return `false`.

Done.

> Could you also make it work with ASSA tags, like `"{\\i1}How are you?{\\i0}"` ? :)

Also done plus it will now handle even if the ending also includes the html closing tag e.g: **</font>{\\i0} or {\\i0}</font>**

**Pull request merged**